mkmimo_multithreaded.c

```diff
@@ -1,15 +1,310 @@
 #include "mkmimo_multithreaded.h"
 #include "queue.h"
+#include <pthread.h>
 
-// int initialize_buffers(Input *inputs, Output *outputs, Queue *empty_buffers)
-// {
-//}
+/**
+ * Parameters
+ */
+#define DEFAULT_MULTIBUFFERING 2  // double buffering
+static int MULTIBUFFERING = DEFAULT_MULTIBUFFERING;
 
-int mkmimo_multithreaded(Inputs *inputs, Outputs *outputs) {
-  // Queue *full_buffers = new_queue();
-  // Queue *empty_buffers = new_queue();
+/**
+  * Buffer pools
+  */
+static Queue *full_buffers;
+static Queue *empty_buffers;
 
-  // initialize_buffers();
+/**
+  * Flags
+  */
+static bool data_is_flowing_in = true;
+static bool data_should_flow_in = true;
+static bool data_should_flow_out = true;
 
-  return 1;  // Not implemented yet
+/**
+  * Stop all threads
+  */
+static inline void teardown_all_threads(void) {
+  // XXX this tears down all input threads
+  data_should_flow_in = false;
+  // XXX this tears down all output threads
+  data_should_flow_out = false;
+}
+
+/**
+ * Find the last record separator in a full buffer.
+ */
+static inline void find_record_separator(Buffer *buf,
+                                         int scan_end_of_record_down_to) {
+  for (int j = buf->begin + buf->size - 1; j >= scan_end_of_record_down_to;
+       --j) {
+    char c = ((char *)buf->data)[j];
+    if (c == '\n') {
+      buf->end_of_last_record = j;
+      break;
+    }
+  }
+}
+
+/**
+  * Grab a buffer from the empty pool and clear it for fresh data.
+  */
+static inline Buffer *grab_empty_buffer(void) {
+  Buffer *buf = dequeue_or_wait(empty_buffers);
+  clear_buffer(buf);
+  return buf;
+}
+
+/**
+ * Function executed by the input threads. Grabs an empty buffer from
+ * the empty buffers queue, fills it, and adds it to the full buffers
+ * queue for processing by the output threads.
+ */
+static void *read_buffers_from_input(void *arg) {
+  Input *input = arg;
+
+  input->buffer = grab_empty_buffer();
+  DEBUG("%s: grabbed an empty buffer %p", input->name, input->buffer);
+  while (data_should_flow_in) {
+    // Read from input to fill up the buffer with at least one record
+    Buffer *buf = input->buffer;
+    int scan_end_of_record_down_to = buf->end_of_last_record + 1;
+    for (;;) {
+      int num_bytes_readable = buf->capacity - buf->size;
+      DEBUG("%s: can read %d bytes", input->name, num_bytes_readable);
+
+      int num_bytes_read = read(input->fd, buf->data + buf->begin + buf->size,
+                                num_bytes_readable);
+      DEBUG("%s: %d bytes read", input->name, num_bytes_read);
+
+      if (num_bytes_read < 0) {
+        // Close input upon errors
+        perrorf("read %s", input->name);
```

Comment:
Message is not clear that there is an IO error. Maybe also print out `num_bytes_read` if the code is meaningful.